### ISSUE TITLE
Please add an option to make string search case-insensitive

### ISSUE THREAD

**Issue by Alan47** [thinkaurelius/titan#1112]
As of 0.9.0-M2, Titan offers String search and Text search with the following properties:
- **String search:** case-sensitive, matches entire string, no text analysis
- **Text search:** case-insensitive, matches words, text analysis (stemming, ignoring special characters, ...)

Currently, there is **no way** in Titan to cover the following (very common) scenario:
- Vertices store a "name" property with less than 255 characters
- Search by name
- Case insensitive
- Allow inclusion of special characters (braces, spaces, punctuation...)

Therefore, it is an _essential_ feature to allow string search to be case insensitive. My suggestion here would be to add another text indexing parameter that decides if the index in question is case sensitive or not.

**Comment by dmill-bz** [comment]
or perhaps provide a new Predicate for this scenario? I don't know how feasible this is but some fields like emails need to be case sensitive fields but the application layers need to at least do a case insensitive search on them.

**Comment by mbroecheler** [comment]
Titan distinguishes between TEXT and STRING search:
http://s3.thinkaurelius.com/docs/titan/0.9.0-M2/index-parameters.html

Does that not cover your use case?

**Comment by Alan47** [comment]
@mbroecheler : no, unfortunately TEXT doesn't cover this case for various reasons. Lucene (and, to my knowledge, Elastic search too) analyzes the text it is given and performs a series of simplifications. For example, the text is split at word boundaries, and everything with less than 3 characters is ignored (i.e. cannot be searched for). Furthermore, all non-alphanumeric characters are ignored as well, and things are even worse when the search string contains spaces. All of this is completely reasonable for large bodies of text, but in particular for a "by-name" style search (where the strings are short), a higher level of precision is required. You simply cannot afford characters being ignored or cut away.

Long story short: it's an entirely different thing, and lucene definitly does not cover that scenario, sadly.

**Comment by Alan47** [comment]
Here's a concrete example. Let's consider the following list of strings:
"Italian"
"IT Management"
"Windows Server (Physical)"

Now let's use the search string "it" (note the lower case) doing a contains search. With TEXT, we have to give Lucene a Regex ".*it.*". What we get is:

"Italian"

... we do not see "IT Management", not because of the upper case, but "IT" has less than 3 letters and got deleted. If we run the same query using STRING, we get nothing, because the case does not match.

What happens if we search for "physical" and "(physical)" on the same set? For the first one, using TEXT, we get:

"Windows Server (Physical)"

... which is correct. However, if we use the second search term (including the braces), we get the _empty_ set. The braces were removed in the lucene index, attempting to search for them explicitly yields no result.

Therefore, my request is: Please make it possible to have case-insensitive STRING search in Titan. TEXT search really does not apply here.

**Comment by BrynCooke** [comment]
Isn't this configurable at the elastic search level?
Take a look at:
https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-lang-analyzer.html#english-analyzer

Maybe just using the whitespace tokenizer and lowercase filter will give you what you need.

**Comment by Alan47** [comment]
Maybe this is possible if one has access to the indexer instance used by Titan and can configure it directly. However, I'm not aware of such a feature in the Titan API, please do correct me if I'm wrong here.

**Comment by BrynCooke** [comment]
You could try create.ext?  http://s3.thinkaurelius.com/docs/titan/0.5.4/elasticsearch.html#_embedding_es_index_creation_settings_with_literal_create_ext_literal

**Comment by Alan47** [comment]
@BrynCooke: Thanks for the pointer. I was always looking for Lucene options in the Titan docs, it appeara that Elasticsearch supports a lot more options. Most likely we will switch to ES.

I've been reading the Titan Docs on ES and thea good portion of the ES docs. Yet I can't see how I would add a tokenizer or a filter (e.g. lowercase filter) via Titan's create.ext properties. Could you perhaps give an example? Sorry, I'm new to most of this configuration business.

**Comment by Alan47** [comment]
I tried it with the following titan properties file:

storage.backend=inmemory
index.search.backend=elasticsearch
index.search.elasticsearch.interface=NODE
index.search.elasticsearch.create.ext.analysis.analyzer.myanalyzer.type=custom
index.search.elasticsearch.create.ext.analysis.analyzer.myanalyzer.tokenizer=mytokenizer
index.search.elasticsearch.create.ext.analysis.analyzer.myanalyzer.filter=mylowercasefilter
index.search.elasticsearch.create.ext.analysis.filter.mylowercasefilter.type=lowercase
index.search.elasticsearch.create.ext.analysis.filter.mylowercasefilter.language=greek
index.search.elasticsearch.create.ext.analysis.tokenizer.mytokenizer.type=whitespace
index.search.elasticsearch.create.ext.analysis.tokenizer.mytokenizer.max_token_length=255

... but Titan tells me the following:

java.lang.IllegalArgumentException: Unknown configuration element in namespace [root.index.elasticsearch.create.ext]: analyzer

And similar errors for the other entries. According to the docs, that should work, right? I tried to mimic the structure from the Elasticsearch docs:

https://www.elastic.co/guide/en/elasticsearch/reference/1.3/analysis-lowercase-tokenfilter.html

**Comment by BrynCooke** [comment]
Which version of titan are you using?

**Comment by Alan47** [comment]
I'm currently on 0.9.0-M2. I can give it a try with 0.5.4. or 0.9.0-M1 when I get home today.

Is this the way it is supposed to work, i.e. did I do it right?

**Comment by BrynCooke** [comment]
I _think_ so. Please raise a separate issue for the error you are getting.

**Comment by Alan47** [comment]
I re-checked the error messages and opened a new issue. If the ES configuration through Titan will eventually work, this ticket can be closed as it is possible to achieve the required search through ES.

Thanks again to @BrynCooke for pointing me into the right direction. The new ticket is available here:
https://github.com/thinkaurelius/titan/issues/1114

**Comment by Alan47** [comment]
I've found a solution to this problem that is viable for us. It might not work for very VERY large datasets, but here's what I did, in case somebody else needs this.

1) Create a regular mixed index in Titan with the STRING parameter (not TEXT!)
2) Query this index as follows:

```
g.V().has("indexedStringProperty", Text.REGEX, "(?i)someregex")
```

The query above is case insensitive. Why is that? Titan uses standard `java.util.regex` classes. Which means that the full functionality is available. There are quite a few hidden gems in that API. Among many others, there is a special construct `(?i)` which, when attached to the beginning of a regular expression, causes the resulting `Pattern` object to be case-insensitive.

**Comment by dmill-bz** [comment]
Oh that's a great tip. Thanks.

**Comment by ghost** [comment]
> > Please make it possible to have case-insensitive STRING search in Titan.

Is this now a documentation defect to add information about (?i) into the Titan docs ?
Is there a reason to change Titan code to support this feature differently ?
